### ISSUE TITLE
High memory usage of `linker.evaluation.prediction_errors_from_labels_table`

### ISSUE THREAD

**Issue by RobinL** [moj-analytical-services/splink#2350]
On our internal linking jobs, this has failed on a machine wiht 128Gb of memory, even when running: 

```
compute_tf_tables(linker)
pipeline = CTEPipeline()
compute_df_concat_with_tf(linker, pipeline)
logger = logging.getLogger("data_linking")
logger.info("computed df concat with tf")
lab_as_sdf = linker.table_management.register_labels_table(labels_df)
prediction_errors = linker.evaluation.prediction_errors_from_labels_table(lab_as_sdf)
```

Particularly weird since the 'big' join condition

```sql
select blah
    from
    __splink__labels_prepared_for_joining as df_labels
    inner join __splink__df_concat_with_tf as l
    on l.person_id = df_labels.person_id_l
    inner join __splink__df_concat_with_tf as r
    on r.person_id = df_labels.person_id_r

```

Is essentially the same as that used in `predict()`, which works.

The difference is that the `predict()` one has 

```sql
select blah
   from __splink__df_concat_with_tf as l
    inner join __splink__blocked_id_pairs as b
    on l."unique_id" = b.join_key_l
    inner join __splink__df_concat_with_tf as r
    on r."unique_id" = b.join_key_r
```

But in that case `__splink__blocked_id_pairs` is a much larger table and so should use more memory

 This join is failing on a labels table with only 27,179 labels

This leads me to suspect that it's some other part of the SQL that's causing the problem

Error message is:
```
Error was: Out of Memory Error: Failed to allocate block of 32768 bytes (bad allocation)
```

**Comment by RobinL** [comment]
Here's a test script that monitors memory usage:

<details>

```py
import duckdb
import psutil
import time
import threading
import splink.comparison_library as cl
from splink import DuckDBAPI, Linker, SettingsCreator, block_on, splink_datasets
from splink.datasets import splink_dataset_labels

# Function to monitor memory usage in a separate thread
def monitor_memory_usage(interval=0.05):
    process = psutil.Process()
    while monitor_thread_active:
        memory_info = process.memory_info()
        print(f"Memory usage: {memory_info.rss / (1024 * 1024):.2f} MB")
        time.sleep(interval)

# Start the memory monitoring thread
monitor_thread_active = True
monitor_thread = threading.Thread(target=monitor_memory_usage, daemon=True)
monitor_thread.start()

# Your existing script
con = duckdb.connect()
con.execute("SET memory_limit = '90MB';")
db_api = DuckDBAPI(connection=con)

df = splink_datasets.fake_1000

settings = SettingsCreator(
    link_type="dedupe_only",
    comparisons=[
        cl.JaroWinklerAtThresholds("first_name", [0.9, 0.7]),
        cl.JaroAtThresholds("surname", [0.9, 0.7]),
        cl.DateOfBirthComparison(
            "dob",
            input_is_string=True,
            datetime_metrics=["year", "month"],
            datetime_thresholds=[1, 1],
        ),
        cl.ExactMatch("city").configure(term_frequency_adjustments=True),
        cl.EmailComparison("email"),
    ],
    blocking_rules_to_generate_predictions=[
        block_on("first_name"),
        block_on("surname"),
    ],
    retain_matching_columns=True,
    retain_intermediate_calculation_columns=True,
)

linker = Linker(df, settings, db_api)

# Execute your functions as usual
linker.training.estimate_probability_two_random_records_match(
    [block_on("first_name", "surname")],
    recall=0.7,
)

linker.training.estimate_u_using_random_sampling(max_pairs=1e6)

linker.training.estimate_parameters_using_expectation_maximisation(
    block_on("first_name", "surname")
)

linker.training.estimate_parameters_using_expectation_maximisation(block_on("dob"))

pairwise_predictions = linker.inference.predict(threshold_match_weight=-10)

df_labels = splink_dataset_labels.fake_1000_labels
labels_table = linker.table_management.register_labels_table(df_labels)

splink_df = linker.evaluation.prediction_errors_from_labels_table(
    labels_table, include_false_negatives=True, include_false_positives=False
)
splink_df.as_pandas_dataframe()

# Stop the memory monitoring thread
monitor_thread_active = False
monitor_thread.join()

print("Memory monitoring completed.")

```
</details>

**Comment by RobinL** [comment]
This seems to be fixed by upgrading to `pip install duckdb==1.0.1.dev4341` from 1.0.0
